### PR TITLE
Remove thread macros from the loader.

### PR DESCRIPTION
**Pull request by davidchisnall**
The loader was initialising threads given two (fairly complex) macros passed via `-D` at compile time.  This caused two problems for auditing:

 - The loader .o files were different depending on thread configuration.
 - The description of threads was interleaved in the loader and hard to separate.

In the new structure, the unlinked loader binaries should be identical for all firmware images.  After linking, they will embed the address of the image header (which should make the hashes of the linked version predictable by anyone who has a copy of the unlinked binary).  All of the thread information now lives in the .thread_config section and so is amenable to being added to the linker report (not done yet).

This also removes the bump allocator from the loader.  This caused some problems with certain stack sizes because it didn't really understand representability.  The build system now puts each stack and trusted stack in a separate section and uses the `CAPALIGN` directive to ensure that they're padded sufficiently.  In a future version, it might sort these by size to ensure that they're in the most efficient order to minimise padding (though in the common case, none of them require padding).

### CODE REVIEW

File: sdk/core/loader/boot.S [hunk 1]
```diff
@@ -89,8 +89,19 @@ start:
 	cspecialrw		ca3, mscratchc, ca3
 	// ca4 still has the RW memory root; nothing to change
 	// The return value is SchedEntryInfo.
-	cincoffset		csp, csp, -16 - CONFIG_THREADS_NUM * BOOT_THREADINFO_SZ
-	csetbounds		ca0, csp, 16 + CONFIG_THREADS_NUM * BOOT_THREADINFO_SZ
+
+	// Calculate space for the structure passed from the loader to the
+	// scheduler.
```
Comment (nwf-msr) [inline sdk/core/loader/boot.S:94]: Assuming it's correct...
```suggestion
	// scheduler.  This is a `class ThreadInfo` from `loader/types.h`.
```

File: compile_flags.txt [hunk 2]
```diff
@@ -31,8 +31,7 @@ riscv32-unknown-unknown
 -DDEVICE_EXISTS_shadow
 -DDEVICE_EXISTS_uart
 -DDEVICE_EXISTS_clint
--DCONFIG_THREADS={{1,2,1024,5},{2,1,1024,5},{3,1,1024,5},}
--DCONFIG_THREADS_ENTRYPOINTS={LA_ABS(__export_test_runner__Z9run_testsv),LA_ABS(__export_thread_pool__Z15thread_pool_runv),LA_ABS(__export_thread_pool__Z15thread_pool_runv),}
+-DCHERIOT_LOADER_TRUSTED_STACK_SIZE=208
```
Comment (nwf-msr) [inline compile_flags.txt:34]: You might explain why 208?  This is also set to 224 down in sdk/xmake.lua?

File: sdk/xmake.lua [hunk 3]
```diff
@@ -352,24 +352,69 @@ rule("firmware")
...
+			"\n\t}\n"
+		local thread_trusted_stack_template =
+			"\n\t. = ALIGN(8);" ..
+			"\n\t.thread_trusted_stack_${thread_id} : CAPALIGN" ..
+			"\n\t{" ..
+			"\n\t\t.thread_${thread_id}_trusted_stack_start = .;" ..
+			"\n\t\t. += ${trusted_stack_size};" ..
+			"\n\t\t.thread_${thread_id}_trusted_stack_end = .;" ..
+			"\n\t}\n"
+		local thread_template =
```
Comment (nwf-msr) [inline sdk/xmake.lua:379]:
```suggestion
		-- Build a `class ThreadConfig` for a thread
		local thread_template =
```

File: sdk/xmake.lua [hunk 4]
```diff
@@ -352,24 +352,69 @@ rule("firmware")
...
 
+		local loader = target:deps()['cheriot.loader'];
+		local loader_stack_size = loader:get('loader_stack_size')
+		local loader_trusted_stack_size = loader:get('loader_trusted_stack_size')
+
 		-- Get the threads config and prepare the predefined macros that describe them
 		local threads = target:values("threads")
+
+
+		local thread_stack_template =
```
Comment (nwf-msr) [inline sdk/xmake.lua:363]:
```suggestion
		-- Declare space and start and end symbols for a thread's C stack
		local thread_stack_template =
```

File: sdk/xmake.lua [hunk 5]
```diff
@@ -352,24 +352,69 @@ rule("firmware")
...
+			"\n\t. = ALIGN(16);" ..
+			"\n\t.thread_stack_${thread_id} : CAPALIGN" ..
+			"\n\t{" ..
+			"\n\t\t.thread_${thread_id}_stack_start = .;" ..
+			"\n\t\t. += ${stack_size};" ..
+			"\n\t\t.thread_${thread_id}_stack_end = .;" ..
+			"\n\t}\n"
+		local thread_trusted_stack_template =
```
Comment (nwf-msr) [inline sdk/xmake.lua:371]:
```suggestion
		-- Declare space and start and end symbols for a thread's trusted stack
		local thread_trusted_stack_template =
```

File: sdk/firmware.ldscript.in [hunk 6]
```diff
@@ -192,12 +193,23 @@ SECTIONS
 		SHORT(@library_count@);
 		# Number of compartment headers.
 		SHORT(@compartment_count@);
-
 		@compartment_headers@
+	}
 
+	# Thread configuration.  This follows the compartment headers but is in a
+	# separate section to make auditing easier.
```
Comment (nwf-msr) [inline sdk/firmware.ldscript.in:200]:
```suggestion
	# separate section to make auditing easier.
	# This section holds a `class ThreadInfo` (loader/types.h)
```

File: sdk/core/loader/boot.cc [hunk 7]
```diff
@@ -1004,6 +933,17 @@ extern "C" SchedulerEntryInfo loader_entry_point(const ImgHdr &imgHdr,
                                                  void         *almightySeal,
                                                  void         *almightyRW)
 {
+	// This relies on a slightly surprising combination of two C++ features:
```
Comment (nwf-msr) [inline sdk/core/loader/boot.cc:936]: Can you expand on what "This" is?  I'm sure it's quite informative, but this comment (or, at least, the `SchedulerEntryInfo` type) seems unrelated to the rest of the PR?

File: sdk/core/loader/boot.S [hunk 8]
```diff
@@ -89,8 +89,19 @@ start:
...
+	la_abs			s1, __thread_count
+	csetaddr		cs1, ca4, s1
+	clhu			s1, 0(cs1)
+	li			t0, -BOOT_THREADINFO_SZ
+	mul			s1, s1, t0
+	addi			s1, s1, -16
```
Comment (nwf-msr) [inline sdk/core/loader/boot.S:100]: What's 16, here?

File: sdk/core/scheduler/main.cc [hunk 9]
```diff
@@ -124,8 +124,10 @@ namespace sched
 
 		for (size_t i = 0; auto *threadSpace : threadSpaces)
 		{
+			Debug::log("Created thread for trusted stack {}",
+			           info[i].trustedStack);
 			Thread *th = new (threadSpace)
-			  Thread(info[i].trustedStack, info[i].threadid, info[i].priority);
+			  Thread(info[i].trustedStack, i + 1, info[i].priority);
```
Comment (nwf-msr) [inline sdk/core/scheduler/main.cc:130]: This seems a little strange to me, since it means that the small-integer thread IDs are now assigned by link order (and will eventually be assigned by the outcome of sorting stacks for capalign packing), right?  If they show up in logging output, their mysterious sourcing might be unsettling?  (I don't think they show up in APIs, tho', so at least there's that!)

File: sdk/core/loader/boot.S [hunk 10]
```diff
@@ -89,8 +89,19 @@ start:
 	cspecialrw		ca3, mscratchc, ca3
 	// ca4 still has the RW memory root; nothing to change
 	// The return value is SchedEntryInfo.
```
Comment (davidchisnall) [inline sdk/core/loader/boot.S:91]:
```suggest
	// The return value is SchedEntryInfo in ca0 (see comment in loader entry).
```

File: sdk/core/loader/boot.S [hunk 11]
```diff
@@ -89,8 +89,19 @@ start:
...
-	cincoffset		csp, csp, -16 - CONFIG_THREADS_NUM * BOOT_THREADINFO_SZ
-	csetbounds		ca0, csp, 16 + CONFIG_THREADS_NUM * BOOT_THREADINFO_SZ
+
+	// Calculate space for the structure passed from the loader to the
+	// scheduler.  This is a `class ThreadInfo` from `loader/types.h`.
```
Comment (davidchisnall) [inline sdk/core/loader/boot.S:94]: Ooops, I shouldn't have committed this proposed change.  This is actually a 	`SchedulerEntryInfo` from `boot.cc`.

```suggestion
```

File: sdk/core/loader/boot.S (hunk 8 above)
Comment (nwf-msr) [reply sdk/core/loader/boot.S:99]: Can you make that an `EXPORT_ASSEMBLY_OFFSET` rather than `16`?

File: sdk/core/loader/boot.S [hunk 12]
```diff
@@ -126,7 +136,8 @@ start:
 	// All other registers will be cleared in the clear-regs block
 
 	// Pass the array of threadInfos as first argument.
-	csetbounds		ca0, csp, CONFIG_THREADS_NUM * BOOT_THREADINFO_SZ
+	addi			s1, s1, -16
```
Comment (nwf-msr) [inline sdk/core/loader/boot.S:139]: This one, too, could benefit from being an `EXPORT_ASSEMBLY_OFFSET`.

**Review by nwf-msr — APPROVED** [review]
Some nits around numeric constants, but otherwise LGTM.  Thanks for the explanation, earlier.

File: compile_flags.txt (hunk 2 above)
Comment (davidchisnall) [reply compile_flags.txt:34]: This file can't contain comments and is used only by CI, so I'm not that worried.

File: sdk/core/loader/boot.S (hunk 12 above)
Comment (davidchisnall) [reply sdk/core/loader/boot.S:139]: These probably should be, I started that refactoring but it's probably better to finish in a separate NFC PR.

**Review by rmn30 — APPROVED** [review]
Looks OK. Bit concerned about the possibly undefined / non-portable behaviour. More / better docs and comments would be good.

**Comment by davidchisnall** [comment]
> Looks OK. Bit concerned about the possibly undefined / non-portable behaviour. More / better docs and comments would be good.

It's a standard C++ feature composed with a standard C feature that every C++ compiler must support if it wants to have a chance of parsing C header files, so I'm fairly confident that it will work anywhere.
